### PR TITLE
HELM 386:Chart Verifier api integration

### PR DESCRIPTION
**Pull request by Kartikey-star**
This is the chart verifier implementation with ODC

Signed-off-by: Kartikey Mamgain <mamgainkartikey@gmail.com>

### CODE REVIEW

File: docs/helm/helm-chart-verification-via-console.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Helm Chart Verification
+
+Chart Verifier is a cli tool that runs a series of independent checks defined in different profiles. This CLI tool is used on our Chart Certification Flow.
```
Comment (Srivaralakshmi) [inline docs/helm/helm-chart-verification-via-console.md:3]:
```suggestion
Chart-verifier is a CLI tool that runs a series of independent checks defined in different profiles. This CLI tool is used to validate the Helm chart against a configurable list of checks through an automated Red Hat OpenShift certification workflow. Certification assures the integrity of the chart and ensures that the Helm chart works seamlessly on Red Hat OpenShift clusters.
```

File: docs/helm/helm-chart-verification-via-console.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+The Helm Chart Verifier api `/api/helm/verify` is an integration of this cli tool  on `Openshift Developer Console`(ODC) and this will provide user more insight on whether the helm charts they are targeting will run well on OpenShift. The following checks will be run when user is looking to verify their helm chart from console:
```
Comment (Srivaralakshmi) [inline docs/helm/helm-chart-verification-via-console.md:5]:
```suggestion
The `/api/helm/verify` Helm chart-verifier API integrates the chart-verifier CLI tool on Openshift Developer Console (ODC). This API provides users more insight into whether the Helm charts that they are targeting will run well on OpenShift. The following checks are run when the user needs to verify their Helm chart from the console:
```

File: docs/helm/helm-chart-verification-via-console.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+3. ContainsValues: Checks if values file exist
+4. HasKubeVersion: Checks kubernetes version specified
+5. HasReadme: Checks chart has a README
+6. HelmLint: checks Helm lint for the chart
+7. ImagesAreCertified: Checks if image is Red Hat certified 
+8. IsHelmV3: Checks the apiversion for the chart and whether it is used in Helm 3
+9. NotContainCsiObjects: Checks whether CSI objects are present
+10. NotContainsCRDs: Checks chart does not contain CRDs
+11. RequiredAnnotationsPresent: Checks all required annotations are present
+
```
Comment (Srivaralakshmi) [inline docs/helm/helm-chart-verification-via-console.md:18]: Suggestion: Put in table format with headers as Checks and Description (2 columns)

@Kartikey-star We Will sit together to review this tomorrow when you are available.

File: docs/helm/helm-chart-verification-via-console.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+As you can see a chart which goes through the certification process verifies that the  Helm chart is compliant with these set of independent checks and is recommended to be used.
```
Comment (Srivaralakshmi) [inline docs/helm/helm-chart-verification-via-console.md:19]:
```suggestion
A chart that goes through the certification process verifies that the Helm chart is compliant with the predefined set of independent checks and is recommended for use.
```

File: docs/helm/helm-chart-verification-via-console.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+The output of the chart verifier api is JSON encoded [Release structure](https://github.com/redhat-certification/chart-verifier/blob/main/pkg/chartverifier/report/types.go#L12-18)
```
Comment (Srivaralakshmi) [inline docs/helm/helm-chart-verification-via-console.md:21]:
```suggestion
The output of the char-verifier API is JSON encoded [Release structure](https://github.com/redhat-certification/chart-verifier/blob/main/pkg/chartverifier/report/types.go#L12-18).
```
Suggestion: Probably needs a rewrite for more clarity.

File: docs/helm/helm-chart-verification-via-console.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+9. NotContainCsiObjects: Checks whether CSI objects are present
+10. NotContainsCRDs: Checks chart does not contain CRDs
+11. RequiredAnnotationsPresent: Checks all required annotations are present
+
+As you can see a chart which goes through the certification process verifies that the  Helm chart is compliant with these set of independent checks and is recommended to be used.
+
+The output of the chart verifier api is JSON encoded [Release structure](https://github.com/redhat-certification/chart-verifier/blob/main/pkg/chartverifier/report/types.go#L12-18) 
+
+## Additional Resources
+To read more about `Chart Verifier` please go through  [RedHat Chart Verifier](https://github.com/redhat-certification/chart-verifier).
```
Comment (Srivaralakshmi) [inline docs/helm/helm-chart-verification-via-console.md:24]:
```suggestion
For more information on `chart-verifier`, see [RedHat Chart Verifier](https://github.com/redhat-certification/chart-verifier).
```

**Review by Srivaralakshmi — CHANGES_REQUESTED** [review]
@Kartikey-star Provided some suggestions. PTAL and address them. We will sit together to review the list of checks tomorrow when you are available.

File: docs/helm/helm-chart-verification-via-console.md (hunk 3 above)
Comment (Srivaralakshmi) [reply docs/helm/helm-chart-verification-via-console.md:18]: @Kartikey-star Please follow the descriptions written here: https://github.com/openshift/console/pull/11964/files#r962695319

**Review by Srivaralakshmi — APPROVED** [review]
@Kartikey-star LGTM. Thanks for the changes!

File: docs/helm/helm-chart-verification-via-console.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+| ContainsTest | Checks that the Helm chart contains at least one test file. |
+| HasKubeVersion | Checks that the `Chart.yaml` file of the Helm chart includes the `kubeVersion` field (v1.0) and is a valid semantic version (v1.1). |
+| ContainsValuesSchema | Checks that the Helm chart contains a JSON schema file (`values.schema.json`) to validate the `values.yaml` file in the chart. |
+| NotContainsCRDs | Checks that the Helm chart does not include custom resource definitions (CRDs). |
+| NotContainCsiObjects |  Checks that the Helm chart does not include Container Storage Interface (CSI) objects. |
+| ImagesAreCertified | Checks that the images referenced by the Helm chart are Red Hat-certified .|
+|  HelmLint | Checks that the chart is well formed by running the `helm lint` command. |
+| ContainsValues | Checks that the Helm chart contains the `values`[¹](https://github.com/redhat-certification/chart-verifier/blob/main/docs/helm-chart-checks.md#-for-more-information-on-the-values-file-see-values-and-best-practices-for-using-values) file. |
+| RequiredAnnotationsPresent |Checks that the Helm chart contains the annotation: ```charts.openshift.io/name```. |
```
Comment (dperaza4dustbit) [inline docs/helm/helm-chart-verification-via-console.md:38]: Is there a chart already in chart verifier docs we can reference?

File: docs/helm/helm-chart-verification-via-console.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+A chart that goes through the certification process verifies that the Helm chart is compliant with the predefined set of independent checks and is recommended for use.
+
+The output of the chart-verifier API is a JSON-encoded [response ](https://github.com/redhat-certification/chart-verifier/blob/main/pkg/chartverifier/report/types.go#L12-18).
```
Comment (dperaza4dustbit) [inline docs/helm/helm-chart-verification-via-console.md:42]: I think we should talk about profiles here, I would create a profile for console under chart verifier. We default to that if api client does not specify the profile.

File: docs/helm/helm-chart-verification-via-console.md (hunk 7 above)
Comment (Kartikey-star) [reply docs/helm/helm-chart-verification-via-console.md:38]: You mean to refer a chart which passes all the checks?

**Review by dperaza4dustbit — APPROVED** [review]
lgtm

File: pkg/helm/handlers/handlerChartVerifier.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	conf := h.getActionConfigurations(h.ApiServerHost, "default", user.Token, &h.Transport)
+	resp, err := h.chartVerifier(req.ChartUrl, req.Values, req.Chartset, req.CheckContainsTest, conf)
+	if err != nil {
+		serverutils.SendResponse(w, http.StatusBadGateway, serverutils.ApiError{Err: fmt.Sprintf("Failed to verify chart: %v", err)})
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	res, err := json.Marshal(resp)
+	if err != nil {
+		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: fmt.Sprintf("Failed to verify chart: %v", err)})
```
Comment (jhadvig) [inline pkg/helm/handlers/handlerChartVerifier.go:68]:
```suggestion
		serverutils.SendResponse(w, http.StatusInternalServerError, serverutils.ApiError{Err: fmt.Sprintf("Failed to encoded response: %v", err)})
```

**Review by jhadvig — COMMENTED** [review]
PR looks good. Adding a few nit comments.

File: pkg/helm/chartverifier/chart_verifier.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	apiVerifier.SetValues(verifier.CommandSet, commandSet)
+	if val, ok := values["openshift-version"]; ok {
+		var value []string
+		value = append(value, val.(string))
+		apiVerifier.SetString(verifier.OpenshiftVersion, value)
+	}
+	if checkContainsTest == false {
+		disabledChecks = append(disabledChecks, apichecks.ContainsTest)
+	}
+	verifier, verifierErr := apiVerifier.UnEnableChecks(disabledChecks).Run(charturl)
```
[the remark this answers is not in the document]
Comment (Kartikey-star) [reply pkg/helm/chartverifier/chart_verifier.go:40]: `UnEnableChecks` is a function in apiVerifier library.

**Review by jhadvig — APPROVED** [review]
/lgtm
/approve

**Comment by Kartikey-star** [comment]
Hi @sanketpathak the api response via postman testing.
Screenshot 2022-09-22 at 3 11 06 PM

**Comment by Kartikey-star** [comment]
Need to wait for https://github.com/redhat-certification/chart-verifier/pull/280 Merge before proceeding with merging this PR.

**Comment by Kartikey-star** [comment]
Hi @sanketpathak the api response got changed . This is the screenshot:
Screenshot 2022-09-23 at 5 52 26 PM

**Review by dperaza4dustbit — APPROVED** [review]
/lgtm

**Pull request merged**